book/src/read-only-read-write-accounts.md

```diff
@@ -0,0 +1,95 @@
+# Read-Only/Read-Write Transaction Accounts
+
+This design covers handling of read-only and read-write accounts in the
+[runtime](runtime.md).  Accounts already distinguish themselves as read-only or
+read-write based on the instruction program.  But to identify read-only accounts
+by program id would require the Account to be fetched and loaded from disk.
+This operation is expensive and while it is occurring the runtime would have to
+reject any transactions referencing the same account.
+
+The proposal introduces a `read_only_accounts` field to the Transaction
+structure and removes the `program_ids` dedicated vector for program accounts.
+
+This design doesn't change the runtime transaction processing rules.
+Programs still can't write or spend accounts that they do not own.  But it
+allows the runtime to optimistically take the right lock for each account
+specified in the transaction before loading the accounts from storage.
+
+## Runtime handling
+
+Read-only accounts have the following properties:
+
+* Can be deposited into.  Deposits can be implemented as a simple `atomic_add`.
+
+* Read-only access to userdata.
+
+Instructions that debit or modify the read only account userdata will fail.
+
+## Account Lock Optimizations
+
+The Accounts module that keeps track of current locked accounts in the runtime
+separates read-only accounts from the read-write accounts.  The read-only
+accounts can be cached in memory and shared between all the threads executing
+transactions.
+
+The current runtime cannot predict if an account is read-only or read-write
```

Comment:
Not following this, are you referring to the program's entry point?  If so, by the time we call the program's entry point the accounts have been loaded and we know definitely if the account should be treated as RO or RW.

Reply:
@jackcmay before the fetch from storage.

Reply:
What entry point are you referring to?  The current wording to me implies you are talking about program entry point,  at that time we do know what locks we need since we've already pulled the account from storage.